    arm64: Correctly align the SVE signal context
    
    The SVE signal context needs to be correctly aligned. Fix this by
    creating a new macro to calculate the needed size to provide this
    alignment, and use it when setting and checking the saved SVE signal
    context.
    
    Approved by:    so
    Security:       FreeBSD-EN-26:02.arm64
    Reported by:    cperciva
    Reviewed by:    cperciva, markj
    Sponsored by:   Arm Ltd
    Differential Revision:  https://reviews.freebsd.org/D54396
    
    (cherry picked from commit a9e77eb7016df70723c208fc09fbd01ec23a732d)
    (cherry picked from commit 683decf362ce0bbfd9ff917618f3e181bc8f1cd0)
--- sys/arm64/arm64/exec_machdep.c.orig
+++ sys/arm64/arm64/exec_machdep.c
@@ -60,6 +60,10 @@
 #include <machine/vfp.h>
 #endif
 
+#define	CTX_SIZE_SVE(buf_size)					\
+    roundup2(sizeof(struct sve_context) + (buf_size),		\
+      _Alignof(struct sve_context))
+
 _Static_assert(sizeof(mcontext_t) == 880, "mcontext_t size incorrect");
 _Static_assert(sizeof(ucontext_t) == 960, "ucontext_t size incorrect");
 _Static_assert(sizeof(siginfo_t) == 80, "siginfo_t size incorrect");
@@ -585,8 +589,7 @@
 
 				buf_size = sve_buf_size(td);
 				/* Check the size is valid */
-				if (ctx.ctx_size !=
-				    (sizeof(sve_ctx) + buf_size))
+				if (ctx.ctx_size != CTX_SIZE_SVE(buf_size))
 					return (EINVAL);
 
 				memset(pcb->pcb_svesaved, 0,
@@ -729,7 +732,7 @@
 {
 	struct sve_context ctx;
 	struct pcb *pcb;
-	size_t buf_size;
+	size_t buf_size, ctx_size;
 	vm_offset_t ctx_addr;
 
 	pcb = td->td_pcb;
@@ -740,14 +743,15 @@
 	MPASS(pcb->pcb_svesaved != NULL);
 
 	buf_size = sve_buf_size(td);
+	ctx_size = CTX_SIZE_SVE(buf_size);
 
 	/* Address for the full context */
-	*addrp -= sizeof(ctx) + buf_size;
+	*addrp -= ctx_size;
 	ctx_addr = *addrp;
 
 	memset(&ctx, 0, sizeof(ctx));
 	ctx.sve_ctx.ctx_id = ARM64_CTX_SVE;
-	ctx.sve_ctx.ctx_size = sizeof(ctx) + buf_size;
+	ctx.sve_ctx.ctx_size = ctx_size;
 	ctx.sve_vector_len = pcb->pcb_sve_len;
 	ctx.sve_flags = 0;
 
